### ISSUE TITLE
Bring in a "require-*"/"no-*" kind of rule

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [JoshuaKGoldberg/eslint-plugin-package-json#51]
https://npmpackagejsonlint.org/docs/rules#require-node-rules has a bunch of `require-*` rules for requiring properties to exist, such as [`require-author`](https://npmpackagejsonlint.org/docs/rules/required-node/require-author), [`require-bin`](https://npmpackagejsonlint.org/docs/rules/required-node/require-bin).

This seems like it'd do well as a general-purpose ESLint rule. It could:

* By default: enforce the required properties that will prompt a warning if missing
* By configuration: be extended to any arbitrary property

Maybe, `package-json/require-properties` or `package-json/require-package-properties` as a name? The latter seems a little overly specific, so I'm leaning towards `require-properties` maybe?

But, what about https://npmpackagejsonlint.org/docs/rules/#disallowed-node-rules? Should we unify into one big rule for requiring which properties do or don't exist?

I also think that https://npmpackagejsonlint.org/docs/rules/required-node/require-scripts can be treated as a subset of https://npmpackagejsonlint.org/docs/rules/scripts/prefer-scripts for this rule. As in, this same rule can require both that `scripts` exists _and_ what values it contains. Someone please yell at me if that's not ideal.

~Blocked on #40, but once that PR is merged this will be good to go.~ ✅

**Comment by JoshuaKGoldberg** [comment]
Yup! You'd have to configure the list yourself, as what constitutes as "useless" changes depending on the context.

**Comment by michaelfaith** [comment]
Maybe one rule called `restrict-properties`?  And it has an array for `require` and `disallow`?  Something like that?

**Comment by michaelfaith** [comment]
What criteria do we use for whether a property should have its own rule vs be configurable in the single umbrella rule?  It feels a little odd to have some properties be broken out as dedicated rules while others are configured as part of one big rule.   The importance of a property is such a relative thing, where to draw that line becomes challenging.

**Comment by JoshuaKGoldberg** [comment]
Agreed, having some be granular and some not would be confusing. Proposal: how about we put _all_ properties in their own rules? That's my preference more and more.

**Comment by michaelfaith** [comment]
> Proposal: how about we put _all_ properties in their own rules? That's my preference more and more.

That sounds good to me.  Then there's the other part of the question: should there be separate `require-` and `validate-` sets of rules.   Personally, I think it makes sense to have those be different things.  Taking https://github.com/JoshuaKGoldberg/eslint-plugin-package-json/pull/786 as an example, having format validation as part of a rule called `require-license` doesn't seem to line up.  So require- rules would just be "does this thing exist or not".  Would `valid-` check that the property exists _and_ check that it's valid, or just checking that if it does exist, it's valid?

**Comment by JoshuaKGoldberg** [comment]
> would `valid-` check...

I think the latter, just checking that if it does exist, it's valid. 👍 otherwise they duplicate the `require-` rules.

**Comment by michaelfaith** [comment]
I think that's right. I can definitely see a case where someone would want to use `valid` without `require`.  But I guess checking my original gut feel: is there a case where someone would want to use `require` _without_ `valid`? That could possibly make a case for `require` to do both after all, and `valid` is just a less strict version of `require`?  Sorry, just thinking out loud. Bikeshedding 😅

**Comment by JoshuaKGoldberg** [comment]
💯 🤝 

We're in alignment. This is great! Looking forward to making progress on this.

**Comment by JoshuaKGoldberg** [comment]
OK! I split out a few dozen `require-` and `valid-` issues -795 through 843- based on the table in #42. There are almost certainly some missing from user error on my end and/or from `package.json` fields that were added after other plugins created their rules. I'm sure we'll discover at the latest when we split apart the `valid-package-def` rule.

Closing this issue out as resolved discussion. Cheers!
